src/tribol/physics/Mortar.cpp

```diff
@@ -631,6 +641,565 @@ void ComputeSingleMortarJacobian( SurfaceContactElem& elem )
+    for ( int i{ 0 }; i < size1; ++i ) {
+      int node_id = mesh1.getGlobalNodeId( elem1, i );
+      for ( int d{ 0 }; d < 3; ++d ) {
+        x1[d * size1 + i] = mesh1.getPosition()[d][node_id];
+        n1[d * size1 + i] = mesh1.getNodalNormals()( d, node_id );
+        f1[d * size1 + i] = 0.0;
+      }
+      p1[i] = mesh1.getNodalFields().m_node_pressure[node_id];
+      g1[i] = 0.0;
+    }
+    int elem2 = plane.getCpElementId1();  // switched from tribol convention
+    RealT x2[12];
+    RealT f2[12];
+    for ( int i{ 0 }; i < size2; ++i ) {
+      int node_id = mesh2.getGlobalNodeId( elem2, i );
+      for ( int d{ 0 }; d < 3; ++d ) {
+        x2[d * size2 + i] = mesh2.getPosition()[d][node_id];
+        f2[d * size2 + i] = 0.0;
+      }
+    }
+    if ( lm_opts.eval_mode == ImplicitEvalMode::MORTAR_RESIDUAL_JACOBIAN ||
+         lm_opts.eval_mode == ImplicitEvalMode::MORTAR_JACOBIAN ) {
+      StackArray<DeviceArray2D<RealT>, 9> blockJ_n( 3 );
+      constexpr int n_disp = 12;
+      for ( int i{ 0 }; i < 2; ++i ) {
+        blockJ_n( i, 0 ) = DeviceArray2D<RealT>( n_disp, n_disp );
+        blockJ_n( i, 0 ).fill( 0.0 );
+      }
+      constexpr int n_multipliers = 4;
+      blockJ_n( 2, 0 ) = DeviceArray2D<RealT>( n_multipliers, n_disp );
+      blockJ_n( 2, 0 ).fill( 0.0 );
+
+      StackArray<DeviceArray2D<RealT>, 9> blockJ( 3 );
+      for ( int i{}; i < 2; ++i ) {
+        for ( int j{}; j < 2; ++j ) {
+          blockJ( i, j ) = DeviceArray2D<RealT>( n_disp, n_disp );
+          blockJ( i, j ).fill( 0.0 );
+        }
+      }
+      for ( int i{}; i < 2; ++i ) {
+        blockJ( i, 2 ) = DeviceArray2D<RealT>( n_disp, n_multipliers );
+        blockJ( i, 2 ).fill( 0.0 );
+        // transpose
+        blockJ( 2, i ) = DeviceArray2D<RealT>( n_multipliers, n_disp );
+        blockJ( 2, i ).fill( 0.0 );
+      }
+      blockJ( 2, 2 ) = DeviceArray2D<RealT>( n_multipliers, n_multipliers );
+      blockJ( 2, 2 ).fill( 0.0 );
+
+      ComputeMortarJacobianEnzyme( x1, n1, p1, f1, blockJ( 0, 0 ).data(), blockJ( 0, 1 ).data(),
+                                   blockJ_n( 0, 0 ).data(), blockJ( 0, 2 ).data(), g1, blockJ( 2, 0 ).data(),
+                                   blockJ( 2, 1 ).data(), blockJ_n( 2, 0 ).data(), size1, x2, f2, blockJ( 1, 0 ).data(),
+                                   blockJ( 1, 1 ).data(), blockJ_n( 1, 0 ).data(), blockJ( 1, 2 ).data(), size2 );
+
+      if ( lm_opts.sparse_mode == SparseMode::MFEM_ELEMENT_DENSE ) {
+        cs->getMethodData()->storeElemBlockJ( { elem1, elem2, elem1 }, blockJ );
+        cs->getdnMethodData()->storeElemBlockJ( { elem1, elem2, elem1 }, blockJ_n );
+      } else {
+        SLIC_WARNING( "Unsupported Jacobian storage method." );
+        return 1;
+      }
+    } else if ( lm_opts.eval_mode == ImplicitEvalMode::MORTAR_GAP ||
+                lm_opts.eval_mode == ImplicitEvalMode::MORTAR_RESIDUAL ) {
+      ComputeMortarForceEnzyme( x1, n1, p1, f1, g1, size1, x2, f2, size2 );
```

Comment:
Lacking some familiarity, I would have assumed the force would have to be computed before the Jacobian. This makes it look like enzyme doesn't need the force, but is responsible for computing df/dx etc. Can you clarify?

Reply:
With Enzyme (and AD in general), you sort of get the force for free (barring any optimizations that could occur if you explicitly say you don't need the force) when you compute the derivative.